File: 587.安装栅栏.py
#
# @lc app=leetcode.cn id=587 lang=python3
#
# [587] 安装栅栏
#

# @lc code=start
class Solution:
    def outerTrees(self, trees: List[List[int]]) -> List[List[int]]:
        n = len(trees)
        if n < 3: return trees
        trees.sort(key = lambda x:(x[0],x[1]))
        cross = lambda a,b,c:(b[0]-a[0])*(c[1]-b[1])-(b[1]-a[1])*(c[0]-b[0])        
        lower = []
        for p in trees:
            while len(lower) > 1 and cross(lower[-2],lower[-1],p) < 0: lower.pop()
            lower.append((p[0],p[1]))        
        upper = []
        for p in reversed(trees):
            while len(upper) > 1 and cross(upper[-2],upper[-1],p) < 0: upper.pop()
            upper.append((p[0],p[1]))            
        return list(set(lower[:-1] + upper[:-1]))
# @lc code=end

File: 587.安装栅栏.go
/*
 * @lc app=leetcode.cn id=587 lang=golang
 *
 * [587] 安装栅栏
 */

// @lc code=start
func cross(p, q, r []int) int {
	return (q[0]-p[0])*(r[1]-q[1]) - (q[1]-p[1])*(r[0]-q[0])
}

func outerTrees(trees [][]int) (ans [][]int) {
	n := len(trees)
	if n < 4 {
		return trees
	}

	leftMost := 0
	for i, tree := range trees {
		if tree[0] < trees[leftMost][0] {
			leftMost = i
		}
	}

	vis := make([]bool, n)
	p := leftMost
	for {
		q := (p + 1) % n
		for r, tree := range trees {
			// 如果 r 在 pq 的右侧，则 q = r
			if cross(trees[p], trees[q], tree) < 0 {
				q = r
			}
		}
		// 是否存在点 i, 使得 p q i 在同一条直线上
		for i, b := range vis {
			if !b && i != p && i != q && cross(trees[p], trees[q], trees[i]) == 0 {
				ans = append(ans, trees[i])
				vis[i] = true
			}
		}
		if !vis[q] {
			ans = append(ans, trees[q])
			vis[q] = true
		}
		p = q
		if p == leftMost {
			break
		}
	}
	return
}

// @lc code=end

